sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -232,16 +231,23 @@ trait GpuHashJoin extends GpuExec {
       GpuColumnVector.from(cb)
     }
 
-    val nvtxRange = new NvtxWithMetrics("hash join", NvtxColor.ORANGE, joinTime)
-    val joined = try {
-      buildSide match {
-        case GpuBuildLeft => doJoinLeftRight(builtTable, streamedTable)
-        case GpuBuildRight => doJoinLeftRight(streamedTable, builtTable)
+    val joined =
+      withResource(new NvtxWithMetrics("hash join", NvtxColor.ORANGE, joinTime)) { _ =>
+        // `doJoinLeftRight` closes the right table if the last argument (`closeRightTable`)
+        // is true, but never closes the left table.
+        buildSide match {
+          case GpuBuildLeft =>
+            // tell `doJoinLeftRight` it is ok to close the `streamedTable`, this can help
+            // in order to close temporary/intermediary data after a filter in some scenarios.
+            doJoinLeftRight(builtTable, streamedTable, true)
```

Comment:
I'm not sure I understand the point of the close boolean argument.  It looks like the table, if specified to close, is always closed after the operation has fully completed, so why not have the caller perform the close?  It would be simpler to read and understand who is responsible for closing, and we could just wrap this in a `withResource(streamedTable)` to make sure streamedTable is always closed as it was in the original code.

Reply:
>  if specified to close, is always closed after the operation has fully completed
Not always. `streamedTable` may be closed by the filter logic (https://github.com/NVIDIA/spark-rapids/pull/1664/files/358f81a81e0c5bab1df51f63738cd864faefe229#diff-753a0d906f1f2d01d8bb08ef13e5205b1710cf915bb0f5fcdd55c3b665d2900dR293). Otherwise, by the time the join executes, we have `streamedTable`, and a filtered version of it + the join. It was an attempt at not having as much in memory around.

We can do it the other way with the `withResource` block here in `doJoin` also. Let me know if the above makes more sense or not.

Reply:
yeah, the desire to free this early is what leads to the messiness.  I'm fine with it since it shouldn't be here long. :smile: